lib/CL/poclcc_binary.c

```diff
@@ -0,0 +1,443 @@
+    {
+      POCL_MEM_FREE(kernel->kernel_name);
+      POCL_MEM_FREE(kernel->binary);
+      POCL_MEM_FREE(kernel->dyn_arguments);
+      POCL_MEM_FREE(kernel->arg_info);
+    }
+}
+
+
+int poclcc_check_global(poclcc_global *binary_format){
+  if (binary_format->version != POCLCC_VERSION)
+    return 0;
+  return !strncmp(binary_format->pocl_id, POCLCC_STRING_ID, POCLCC_STRING_ID_LENGTH);
+}
+
+int poclcc_check_device(poclcc_device *device){
+  if (device->version != POCLCC_VERSION)
+    return 0;
+  return !strncmp(device->pocl_id, POCLCC_STRING_ID, POCLCC_STRING_ID_LENGTH);
+}
+
+#define FNV_OFFSET UINT64_C(0xcbf29ce484222325)
+#define FNV_PRIME UINT64_C(0x100000001b3)
+uint64_t poclcc_get_device_id(cl_device_id device){
+  //FNV-1A with vendor_id, llvm_target_triplet and llvm_cpu
+  uint64_t result = FNV_OFFSET;
+  const char *llvm_tt = device->llvm_target_triplet;
+  const char *llvm_cpu = device->llvm_cpu;
+
+  result *= FNV_PRIME;
+  result ^= device->vendor_id;
+  int i, length = strlen(llvm_tt);
+  for (i=0; i<length; i++)
+    {
+      result *= FNV_PRIME;
+      result ^= llvm_tt[i];
+    }
+  length = strlen(llvm_cpu);
+  for (i=0; i<length; i++)
+    {
+      result *= FNV_PRIME;
+      result ^= llvm_cpu[i];
+    }
+
+  return result;
+}
+
+int poclcc_check_device_id(cl_device_id device, poclcc_device *devicecc){
+  return poclcc_get_device_id(device) == devicecc->device_id;
+}
+
+int poclcc_check_binary(cl_device_id device, const unsigned char *binary){
+  poclcc_device *devicecc = (poclcc_device *)binary;
+  return poclcc_check_device(devicecc) && poclcc_check_device_id(device, devicecc);
+}
+
+
+int poclcc_sizeof_kernel(poclcc_kernel *kernel){
+  return sizeof(poclcc_kernel) 
+    + kernel->sizeofKernelName + kernel->sizeofBinary
+    - sizeof(kernel->kernel_name) - sizeof(kernel->binary) 
+    + (kernel->num_args + kernel->num_locals) * sizeof(struct pocl_argument)
+    + kernel->num_args * sizeof(struct pocl_argument_info)
+    - sizeof (kernel->dyn_arguments) - sizeof(kernel->arg_info);
+}
+
+int poclcc_sizeof_device(poclcc_device *device){
+  int size = sizeof(poclcc_device) - sizeof(device->kernels);
+  int i;
+  for (i=0; i<device->num_kernels; i++)
+    size += poclcc_sizeof_kernel(&(device->kernels[i]));
+
+  return size;
+}
+
+int poclcc_sizeof_global(poclcc_global *binary_format){
+  int size = 0;
+  if (binary_format == NULL)
+    return -1;
+    
+  size += sizeof(poclcc_global) - sizeof(poclcc_device *);
+  int i;
+  for (i=0; i<binary_format->num_devices; i++)
+    size += poclcc_sizeof_device(&(binary_format->devices[i]));
+  
+  return size;
+}
+
+int poclcc_sizeof_global_from_binaries_sizes(size_t *binaries_sizes, int num_devices){
+  if (binaries_sizes == NULL)
+    return -1;
+
+  int i;
+  int size = 0;
+  for (i=0; i<num_devices; i++)
+    size += binaries_sizes[i];
+
+  size += sizeof(poclcc_global) - sizeof(poclcc_device *);
+  return size;
+}
+
+
+void copyKernel2Buffer(poclcc_kernel *kernel, char **buf){
```

Comment:
I'm not sure... This is a translation from the struct (with pointers) to a continuous buffered form of the structure (without pointers).